File: mynlp-fasttext/src/main/java/fasttext/FastText.java

package fasttext;

import com.carrotsearch.hppc.IntArrayList;
import com.google.common.base.Charsets;
import com.google.common.base.Preconditions;
import com.google.common.base.Stopwatch;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.Lists;
import com.google.common.collect.Sets;
import com.google.common.io.Files;
import fasttext.matrix.Matrix;
import fasttext.matrix.Vector;
import fasttext.pq.QMatrix;
import fasttext.utils.*;

import java.io.*;
import java.text.DecimalFormat;
import java.util.*;
import java.util.concurrent.TimeUnit;

import static fasttext.utils.ModelName.sup;

public class FastText {

    final static int FASTTEXT_VERSION = 12;

    final static int FASTTEXT_FILEFORMAT_MAGIC_INT32 = 793712314;

    private Dictionary dict;

    private Matrix input;

    private Matrix output;

    private QMatrix qinput;

    private QMatrix qoutput;

    private Model model;

    private boolean quant = false;//是否量化

    private Args args;

    public FastText(Dictionary dict, Matrix input, Matrix output_, Model model, Args args) {
        this.dict = dict;
        this.input = input;
        this.output = output_;
        this.model = model;
        this.args = args;
    }

    public FastText(Dictionary dict, QMatrix input, QMatrix output_, Model model, Args args) {
        this.dict = dict;
        this.qinput = input;
        this.qoutput = output_;
        this.model = model;
        quant = true;
        this.args = args;
    }


    /**
     * Nearest Neighbor 计算器。可以重复使用
     */
    public NearestNeighbor nearestNeighbor() {
        return new NearestNeighbor();
    }


    public List<FloatStringPair> findNN(Matrix wordVectors, Vector queryVec, int k, Set<String> sets){


        float queryNorm = queryVec.norm();
        if (Math.abs(queryNorm) < 1e-8) {
            queryNorm = 1;
        }

        FloatStringPair[] mostSimilar = new FloatStringPair[k];
        final int mastSimilarLast = mostSimilar.length - 1;
        for (int i = 0; i < mostSimilar.length; i++) {
            mostSimilar[i] = new FloatStringPair(-1f, "");
        }

        //Vector vec = new Vector(args.dim);
        for (int i = 0; i < dict.nwords(); i++) {
            float dp = wordVectors.dotRow(queryVec, i)/queryNorm;
            FloatStringPair last = mostSimilar[mastSimilarLast];
            if (dp > last.first) {
                last.first = dp;
                last.second = dict.getWord(i);
                Arrays.sort(mostSimilar, FloatStringPair.High2Low);
            }
        }

        List<FloatStringPair> result = Lists.newArrayList();
        for (FloatStringPair r : mostSimilar) {
            if (r.first != -1f && !sets.contains(r.second)) {
                result.add(r);
            }
        }
        return result;
    }


    public class NearestNeighbor {

        private Matrix wordVectors;

        public NearestNeighbor() {
            wordVectors = new Matrix(dict.nwords(), args.dim);
            precomputeWordVectors(wordVectors);
        }

        public List<FloatStringPair> nn(String wordQuery, int k) {
            Vector queryVec = getWordVector(wordQuery);
            Set<String> sets = new HashSet<>();
            sets.add(wordQuery);
            return findNN(wordVectors, queryVec, k,sets);
        }


    }

    /**
     * 类推 Query triplet (A - B + C)?
     * 比如 中国 - 北京 + 东京 = 日本
     * 这个计算器需要重用
     */
    public Analogies analogies() {
        return new Analogies();
    }

    public class Analogies{
        private Matrix wordVectors;

        public Analogies() {
            wordVectors = new Matrix(dict.nwords(), args.dim);
            precomputeWordVectors(wordVectors);
        }

        /**
         * Query triplet (A - B + C)?
         * @param A
         * @param B
         * @param C
         * @param k
         */
        public List<FloatStringPair> analogies(String A,String B,String C,int k){
            Vector buffer = new Vector(args.dim);
            Vector query = new Vector(args.dim);

            getWordVector(buffer,A);
            query.add(buffer);

            getWordVector(buffer,B);
            query.add(buffer,-1);

            getWordVector(buffer,C);
            query.add(buffer);

            Set<String> sets = Sets.newHashSet(A, B, C);

            return findNN(wordVectors,query,k,sets);
        }

    }


    /**
     * 计算所有词的向量。
     * 之所以向量都除以norm进行归一化。因为使用者。使用dot表达相似度，也会除以query vector的norm。然后归一化。
     * 最后距离结构都是0 ~ 1 的数字
     * @param wordVectors
     */
    public void precomputeWordVectors(Matrix wordVectors) {
        Vector vec = new Vector(args.dim);
        wordVectors.zero();
        for (int i = 0; i < dict.nwords(); i++) {
            String word = dict.getWord(i);
            getWordVector(vec, word);
            float norm = vec.norm();
            if (norm > 0) {
                wordVectors.addRow(vec, i, 1.0f / norm);
            }
        }
    }

    /**
     * 预测分类标签
     *
     * @param tokens
     * @param k
     * @return
     */
    public List<FloatStringPair> predict(Iterable<String> tokens, int k) {
        IntArrayList words = new IntArrayList();
        IntArrayList labels = new IntArrayList();

        dict.getLine(tokens, words, labels);

        if (words.isEmpty()) {
            return ImmutableList.of();
        }
        Vector hidden = new Vector(args.dim);
        Vector output = new Vector(dict.nlabels());

        List<FloatIntPair> modelPredictions = Lists.newArrayListWithCapacity(k);

        model.predict(words, k, modelPredictions, hidden, output);

        return Lists.transform(modelPredictions, x -> new FloatStringPair(x.first, dict.getLabel(x.second)));
    }


    /**
     * 把词向量填充到一个Vector对象里面去
     *
     * @param vec
     * @param word
     */
    public void getWordVector(Vector vec, final String word) {
        vec.zero();
        final IntArrayList ngrams = dict.getSubwords(word);
        int[] buffer = ngrams.buffer;
        for (int i = 0,len=ngrams.size(); i < len; i++) {
            addInputVector(vec, buffer[i]);
        }

        if (ngrams.size() > 0) {
            vec.mul(1.0f / ngrams.size());
        }
    }

    public Vector getWordVector(String word) {
        Vector vec = new Vector(args.dim);
        getWordVector(vec, word);
        return vec;
    }


    public Vector getSentenceVector(Iterable<String> tokens) {
        Vector svec = new Vector(args.dim);
        getSentenceVector(svec, tokens);
        return svec;
    }

    /**
     * 句子向量
     *
     * @param svec
     * @param tokens
     */
    public void getSentenceVector(Vector svec, Iterable<String> tokens) {
        svec.zero();
        if (args.model == ModelName.sup) {
            IntArrayList line = new IntArrayList();
            IntArrayList labels = new IntArrayList();
            dict.getLine(tokens, line, labels);

            for (int i = 0; i < line.size(); i++) {
                addInputVector(svec, line.get(i));
            }

            if (!line.isEmpty()) {
                svec.mul(1.0f / line.size());
            }
        } else {
            Vector vec = new Vector(args.dim);
            int count = 0;
            for (String word : tokens) {
                getWordVector(vec, word);
                float norm = vec.norm();
                if (norm > 0) {
                    vec.mul(1.0f / norm);
                    svec.add(vec);
                    count++;
                }
            }
            if (count > 0) {
                svec.mul(1.0f / count);
            }
        }
    }

    private void addInputVector(Vector vec, int ind) {
        if (quant) {
            vec.addRow(qinput, ind);
        } else {
            vec.addRow(input, ind);
        }
    }


    /**
     * Load binary model file.
     *
     * @param modelPath
     * @throws IOException
     */
    public static FastText loadModel(String modelPath) throws Exception {
        Stopwatch stopwatch = Stopwatch.createStarted();
        FastText fastText = FastTextIO.readClangModel(modelPath);
        stopwatch.stop();
        System.out.println("load model use time " + stopwatch.elapsed(TimeUnit.MILLISECONDS) + " ms");
        return fastText;
    }

    public void saveModel(File out) throws Exception{
        if (quant) {
            Preconditions.checkArgument(out.getName().endsWith(".ftz"));
        }else{
            Preconditions.checkArgument(out.getName().endsWith(".bin"));
        }

        try (
                BufferedOutputStream bout = new BufferedOutputStream(
                        new FileOutputStream(out), 1024 * 64);
                CLangDataOutputStream dis = new CLangDataOutputStream(bout)) {

            //singModel
            dis.writeInt(FASTTEXT_FILEFORMAT_MAGIC_INT32);
            dis.writeInt(FASTTEXT_VERSION);

            args.save(dis);
            dict.save(dis);

            dis.writeBoolean(quant);
            if (quant) {
                qinput.save(dis);
            }else{
                input.save(dis);
            }

            dis.writeBoolean(args.qout);
            if (quant && args.qout) {
                qoutput.save(dis);
            }else{
                output.save(dis);
            }
        }
    }

    /**
     * 保存词到向量文本
     *
     * @param file
     */
    public void saveVectors(File file) throws Exception {
        Preconditions.checkArgument(file.getName().endsWith(".vec"));
        if (file.exists()) {
            file.delete();
        }
        if (file.getParentFile() != null) {
            file.getParentFile().mkdirs();
        }
        Vector vec = new Vector(args.dim);
        DecimalFormat df = new DecimalFormat("0.#####");

        try (Writer writer = Files.asByteSink(file).asCharSink(Charsets.UTF_8).openBufferedStream()) {
            writer.write(dict.nwords() + " " + args.dim + "\n");
            for (int i = 0; i < dict.nwords(); i++) {
                String word = dict.getWord(i);
                getWordVector(vec, word);
                writer.write(word);
                writer.write(" ");
                for (int j = 0; j < vec.length(); j++) {
                    writer.write(df.format(vec.get(j)));
                    writer.write(" ");
                }
                writer.write("\n");
            }
        }
    }

    /**
     * 分类模型量化
     *
     * @param out
     */
    public void quantize(String out) {
        if (quant) {
            System.out.println("该模型已经被量化过");
            return;
        }
    }


    public static FastText train(File trainFile, ModelName model_name, TrainArgs args) throws Exception {
        return new FastTextTrain().train(trainFile,model_name,args);
    }

    public static FastText train(File trainFile,ModelName model_name) throws Exception {
        return new FastTextTrain().train(trainFile,model_name,new TrainArgs());
    }

    public static FastText train(File trainFile) throws Exception {
        return new FastTextTrain().train(trainFile, ModelName.sup,new TrainArgs());
    }
}

File: fasttext/src/main/java/com/mayabot/nlp/fasttext/FastText.kt
package com.mayabot.nlp.fasttext

import com.mayabot.nlp.fasttext.args.ModelArgs
import com.mayabot.nlp.fasttext.args.ModelName
import com.mayabot.nlp.fasttext.args.TrainArgs
import com.mayabot.nlp.fasttext.blas.*
import com.mayabot.nlp.fasttext.blas.Vector
import com.mayabot.nlp.fasttext.dictionary.Dictionary
import com.mayabot.nlp.fasttext.dictionary.EOS
import com.mayabot.nlp.fasttext.dictionary.buildFromFile
import com.mayabot.nlp.fasttext.loss.createLoss
import com.mayabot.nlp.fasttext.train.FastTextTrain
import com.mayabot.nlp.fasttext.train.FileSampleLineIterable
import com.mayabot.nlp.fasttext.train.SampleLine
import com.mayabot.nlp.fasttext.train.loadPreTrainVectors
import com.mayabot.nlp.fasttext.utils.AutoDataInput
import com.mayabot.nlp.fasttext.utils.IntArrayList
import com.mayabot.nlp.fasttext.utils.loggerln
import com.mayabot.nlp.fasttext.utils.openDataInputStream
import java.io.DataInputStream
import java.io.File
import java.io.IOException
import java.io.InputStream
import java.nio.ByteOrder
import java.text.DecimalFormat
import java.util.*
import kotlin.collections.ArrayList
import kotlin.math.abs
import kotlin.math.exp

class FastText(
        val args: ModelArgs,
        val dict: Dictionary,
        val model: Model,
        val quant: Boolean
) {

    private val input: Matrix = model.wi
    val output: Matrix = model.wo

    /**
     * 预测分类标签
     *
     * @param tokens
     * @param k
     * @return
     */
    fun predict(tokens: List<String>, k: Int, threshold: Float): List<ScoreLabelPair> {

        // 要附加一个EOS标记

        val tokens2 = tokens.toMutableList()
        tokens2.add(EOS)

        val words = IntArrayList()
        val labels = IntArrayList()

        dict.getLine(tokens2, words, labels)

        if (words.isEmpty) {
            return emptyList()
        }

        val result = predict(k, words, threshold)


        return result.map { x -> ScoreLabelPair(exp(x.score), dict.getLabel(x.id)) }
    }

    fun predict(k: Int, words: IntArrayList, threshold: Float): List<ScoreIdPair> {
        if (words.isEmpty) {
            return emptyList()
        }

        if (args.model != ModelName.sup) {
            error("Model needs to be supervised for prediction!")
        }
        val state = Model.State(args.dim, dict.nlabels, 0)
        val predictions = ArrayList<ScoreIdPair>()
        model.predict(words, k, threshold, predictions, state)

        return predictions
    }


    private fun findNN(wordVectors: DenseMatrix, queryVec: Vector, k: Int, sets: Set<String>): List<ScoreLabelPair> {

        var queryNorm = queryVec.norm2()
        if (abs(queryNorm) < 1e-8) {
            queryNorm = 1f
        }

        val mostSimilar = (0 until k).map { ScoreLabelPair(-1f, "") }.toList().toTypedArray()
        val mastSimilarLast = mostSimilar.size - 1

        for (i in 0 until dict.nwords) {
            val dp = wordVectors[i] * queryVec / queryNorm
            val last = mostSimilar[mastSimilarLast]
            if (dp > last.score) {
                last.score = dp
                last.label = dict.getWord(i)

                mostSimilar.sortByDescending { it.score }
            }
        }

        val result = ArrayList<ScoreLabelPair>()
        for (r in mostSimilar) {
            if (r.score != -1f && !sets.contains(r.label)) {
                result.add(r)
            }
        }

        return result
    }

    private val wordVectors: DenseMatrix by lazy {
        /**
         * 计算所有词的向量。
         * 之所以向量都除以norm进行归一化。因为使用者。使用dot表达相似度，也会除以query vector的norm。然后归一化。
         * 最后距离结构都是0 ~ 1 的数字
         * @param wordVectors
         */
        fun preComputeWordVectors(wordVectors: DenseMatrix) {
            val vec = floatArrayVector(args.dim)
            wordVectors.zero()
            for (i in 0 until dict.nwords) {
                val word = dict.getWord(i)
                getWordVector(vec, word)
                val norm = vec.norm2()
                if (norm > 0) {
                    wordVectors[i] += 1.0f / norm to vec
                }
            }
        }

        val matrix = floatArrayMatrix(dict.nwords, args.dim)
        val s1 = System.currentTimeMillis()
        preComputeWordVectors(matrix)
        val s2 = System.currentTimeMillis()
        loggerln("Init wordVectors martix use time ${s2 - s1} ms")
        matrix
    }


    /**
     * NearestNeighbor
     */
    fun nearestNeighbor(wordQuery: String, k: Int): List<ScoreLabelPair> {
        val queryVec = getWordVector(wordQuery)
        val sets = HashSet<String>()
        sets.add(wordQuery)
        return findNN(wordVectors, queryVec, k, sets)
    }

    /**
     * Query triplet (A - B + C)?
     * @param A
     * @param B
     * @param C
     * @param k
     */
    fun analogies(A: String, B: String, C: String, k: Int): List<ScoreLabelPair> {

        val buffer = floatArrayVector(args.dim)
        val query = floatArrayVector(args.dim)

        getWordVector(buffer, A)
        query += buffer

        getWordVector(buffer, B)
        query += -1f to buffer

        getWordVector(buffer, C)
        query += buffer

        val sets = hashSetOf(A, B, C)

        return findNN(wordVectors, query, k, sets)
    }


    /**
     * 把词向量填充到一个Vector对象里面去
     *
     * @param vec
     * @param word
     */
    fun getWordVector(vec: Vector, word: String) {
        vec.zero()
        val ngrams = dict.getSubwords(word)
        val buffer = ngrams.buffer
        var i = 0
        val len = ngrams.size()
        while (i < len) {
            addInputVector(vec, buffer[i])
            i++
        }

        if (ngrams.size() > 0) {
            vec *= 1.0f / ngrams.size()
        }
    }

    fun getWordVector(word: String): Vector {
        val vec = floatArrayVector(args.dim)
        getWordVector(vec, word)
        return vec
    }

    fun test(file: File, k: Int = 1, threshold: Float = 0.0f): Meter {
        return test(FileSampleLineIterable(file),k,threshold)
    }

    fun test(file: Iterable<SampleLine>, k: Int = 1, threshold: Float = 0.0f): Meter {
        val line = IntArrayList()
        val labels = IntArrayList()
        val meter = Meter()
//        val state = Model.State(args.dim,dict.nlabels,0)
        for (sample in file) {
            line.clear()
            labels.clear()
            dict.getLine(sample.words, line, labels)
            if (!labels.isEmpty && !line.isEmpty) {
                val predictions = predict(k, line, threshold)
                meter.log(labels, predictions)

//                if (labels[0] == 1 && predictions[0].id == 1) {
//
//                    line.forEach2 {
//                        loggerln(dict.getWord(it))
//                    }
//                    loggerln("----")
//                }
            }
        }
        meter.print(dict, k, true)
        return meter
    }


    /**
     * 计算句子向量
     * @return 句子向量
     */
    fun getSentenceVector(tokens: Iterable<String>): Vector {
        val svec = floatArrayVector(args.dim)
        getSentenceVector(svec, tokens)
        return svec
    }


    /**
     * 句子向量
     *
     * @param svec
     * @param tokens
     */
    private fun getSentenceVector(svec: Vector, tokens: Iterable<String>) {
        svec.zero()
        if (args.model == ModelName.sup) {
            val line = IntArrayList()
            val labels = IntArrayList()
            dict.getLine(tokens, line, labels)

            for (i in 0 until line.size()) {
                addInputVector(svec, line.get(i))
            }

            if (!line.isEmpty) {
                svec *= (1.0f / line.size())
            }
        } else {
            val vec = floatArrayVector(args.dim)
            var count = 0
            for (word in tokens) {
                getWordVector(vec, word)
                val norm = vec.norm2()
                if (norm > 0) {
                    vec *= (1.0f / norm)
                    svec += vec
                    count++
                }
            }
            if (count > 0) {
                svec *= (1.0f / count)
            }
        }
    }

    private fun addInputVector(vec: Vector, ind: Int) {
        input.addRowToVector(vec, ind)
    }


    /**
     * 把词向量另存为文本格式
     *
     * @param file
     */
    @Throws(Exception::class)
    fun saveVectors(path: String) {
        var fileName = path
        if (!fileName.endsWith("vec")) {
            fileName += ".vec"
        }

        val file = File(fileName).apply {
            if (exists()) delete()
            if (parentFile != null) parentFile.mkdirs()
        }

        val vec = floatArrayVector(args.dim)
        val df = DecimalFormat("0.#####")

        file.bufferedWriter(Charsets.UTF_8).use { writer ->
            writer.write("${dict.nwords} ${args.dim}\n")
            for (i in 0 until dict.nwords) {
                val word = dict.getWord(i)
                getWordVector(vec, word)
                writer.write(word)
                writer.write(" ")
                for (j in 0 until vec.length()) {
                    writer.write(df.format(vec[j].toDouble()))
                    writer.write(" ")
                }
                writer.write("\n")
            }
        }


    }


    /**
     * 保存为自有的文件格式(多文件）
     */
    @Throws(Exception::class)
    fun saveModel(file: String) {
        val path = File(file)
        if (path.exists()) {
            path.deleteRecursively()
        }
        path.mkdirs()

        //dict
        File(path, "dict.bin").outputStream().channel.use {
            dict.save(it)
        }

        //args
        File(path, "args.bin").outputStream().channel.use {
            args.save(it)
        }

        val qInputFlag = if (input is QuantMatrix) "q" else ""
        input.save(File(path, "${qInputFlag}input.matrix"))

        val qOutputFlag = if (output is QuantMatrix) "q" else ""
        output.save(File(path, "${qOutputFlag}output.matrix"))

    }

    /**
     * 对分类模型进行压缩
     */
    @JvmOverloads
    fun quantize(dsub: Int = 2,
//                 cutoff:Int=0,
//                 retrain:Int,
                 qnorm: Boolean = false,
                 qout: Boolean = false): FastText {

//        fun selectEmbedding(): IntArray{
//            val norms = floatArrayVector(input.row)
//            (input as DenseMatrix).l2NormRow(norms)
//            val idx = IntArray(input.row){0}
//            iota(idx)
//            val eosid = dict.getWordId(EOS)
//
//            idx.sortedWith(Comparator { i1, i2 ->
//                if (i1 == eosid && i2 == eosid) {
//                    1
//                }else{
//                    if(eosid == i1){
//                        -1
//                    }else{
//                        if(eosid !=i2){
//                            -(norms[i1].compareTo(i2))
//                        }else{
//                            -1
//                        }
//                    }
//                }
//            })
//
//            return idx.takeLast(idx.size-cutoff).toIntArray()
//        }

        if (this.args.model != ModelName.sup) {
            error("For now we only support quantization of supervised models")
        }

        val normalizeGradient = args.model == ModelName.sup

        // 暂时不实现cutoff
//        if (cutoff > 0 && cutoff < input.row) {
//            val idx = selectEmbedding()
//
//        }

        val input_ = buildQMatrix(this.input as DenseMatrix, dsub, qnorm)

        var output_: Matrix = this.output
        var loss_ = this.model.loss
        if (qout) {
            output_ = buildQMatrix(this.output as DenseMatrix, 2, qnorm)
            loss_ = createLoss(args, output_, args.model, this.dict)
        }

        return FastText(args, dict, Model(input_, output_, loss_, normalizeGradient), true)
    }


    companion object {

        @JvmStatic
        fun trainSupervised(file: File, trainArgs: TrainArgs = TrainArgs()) = train(file, ModelName.sup, trainArgs)

        @JvmStatic
        fun trainCow(file: File, trainArgs: TrainArgs = TrainArgs()) = train(file, ModelName.cbow, trainArgs)

        @JvmStatic
        fun trainSkipgram(file: File, trainArgs: TrainArgs = TrainArgs()) = train(file, ModelName.sg, trainArgs)


        @JvmStatic
        fun train(file: File, modelName: ModelName, trainArgs: TrainArgs): FastText {

            val args = trainArgs.toComputedTrainArgs(modelName)

            fun prepareSources(): List<Iterable<SampleLine>> {
                val parent = FileSampleLineIterable(file)
                var thread = args.thread

                val Size50M = 100 * 1024 * 1024
                var lines = -1
                if (file.length() < Size50M) {
                    lines = parent.lines()

                    // 数量太少
                    if (lines <= thread * 10) {
                        thread = 1
                    }

                }

                return if (thread == 1) {
                    if (file.length() < Size50M) {
                        listOf(parent.toMemList())
                    } else {
                        listOf(parent)
                    }
                } else {
                    if (file.length() < Size50M) {
                        parent.splitMutiFiles(thread).map { it.toMemList() }
                    } else {
                        parent.splitMutiFiles(thread)
                    }
                }

            }

            return train(prepareSources(),modelName,trainArgs)
        }

        @JvmStatic
        fun train(sources: List<Iterable<SampleLine>>, modelName: ModelName, trainArgs: TrainArgs): FastText {

            val args = trainArgs.toComputedTrainArgs(modelName)
            val modelArgs = args.modelArgs

            try {
                val dict = buildFromFile(args, sources, args.maxVocabSize)

                val input = if (args.preTrainedVectors != null) {
                    loadPreTrainVectors(dict, args.preTrainedVectors, args)
                } else {
                    floatArrayMatrix(dict.nwords + modelArgs.bucket, modelArgs.dim)
                            .apply {
                                uniform(1.0f / modelArgs.dim)
                            }
                }

                dict.init()

                val output = floatArrayMatrix(
                        if (ModelName.sup == args.model) dict.nlabels else dict.nwords,
                        modelArgs.dim
                ).apply {
                    zero()
                }

                val loss = createLoss(modelArgs, output, args.model, dict)
                val normalizeGradient = args.model == ModelName.sup

                val model = Model(input, output, loss, normalizeGradient)

                val fastText = FastText(modelArgs, dict, model, false)

                FastTextTrain(args, fastText).startThreads(sources)

                return fastText

            } finally {

                for (source in sources) {
                    if (source is FileSampleLineIterable) {
                        source.file.delete()
                    }
                }
            }
        }

        fun loadCppModel(file: File): FastText {
            return CppFastTextSupport.load(file)
        }

        fun loadCppModel(ins: InputStream): FastText {
            return CppFastTextSupport.loadCModel(ins)
        }

        /**
         * 加载Java模型,[moduleDir]是目录
         */
        fun loadModel(moduleDir: File, mmap: Boolean = false): FastText {

            check(moduleDir.exists() && moduleDir.isDirectory)

            val args = ModelArgs.load(File(moduleDir, "args.bin"))

            val dict = File(moduleDir, "dict.bin").openDataInputStream().use {
                Dictionary.loadModel(args, AutoDataInput(it))
            }

            val quant = File(moduleDir, "qinput.matrix").exists()

            val input = if (quant) {
                loadQuantMatrix(File(moduleDir, "qinput.matrix"))
            } else {
                loadDenseMatrix(File(moduleDir, "input.matrix"), mmap)
            }


            if (!quant && dict.isPruned()) {
                error("Invalid model file.\n" +
                        "Please download the updated model from www.fasttext.cc.\n" +
                        "See issue #332 on Github for more information.\n")
            }

            val output = if (File(moduleDir, "qoutput.matrix").exists()) {
                loadQuantMatrix(File(moduleDir, "qoutput.matrix"))
            } else {
                loadDenseMatrix(File(moduleDir, "output.matrix"), mmap)
            }

            val loss = createLoss(args, output, args.model, dict)

            val normalizeGradient = args.model == ModelName.sup

            return FastText(args, dict, Model(input, output, loss, normalizeGradient), quant)
        }

    }

}

data class ScoreIdPair(val score: Float, val id: Int) {
    override fun toString(): String {
        return "[$id,$score]"
    }
}

data class ScoreLabelPair(var score: Float, var label: String) {
    override fun toString(): String {
        return "[$label,$score]"
    }
}

/**
 * 从C语言版本的FastText产生的模型文件
 */
object CppFastTextSupport {

    /**
     * Load binary model file. 这个二进制版本是C语言版本的模型
     * 从流读取，因为生产环境可能从classpath里面读取模型文件
     * @param input C语言版本的模型的InputStream
     * @return FastTextModel
     * @throws Exception
     */
    @Throws(Exception::class)
    fun loadCModel(inputStream: InputStream): FastText {

        val ins = DataInputStream(inputStream.buffered(1024 * 1024))

        ins.use {
            val buffer = AutoDataInput(it, ByteOrder.LITTLE_ENDIAN)

            //check model
            val magic = buffer.readInt()
            val version = buffer.readInt()

            if (magic != 793712314) {
                throw RuntimeException("Model file has wrong file format!")
            }

            if (version > 12) {
                throw RuntimeException("Model file has wrong file format! version is $version")
            }

            //Args
            val args = run {
                var args_ = ModelArgs.load(buffer)

                if (version == 11 && args_.model == ModelName.sup) {
                    // backward compatibility: old supervised models do not use char ngrams.
                    args_ = args_.copy(maxn = 0)
                }
                args_
            }

            //dictionary
            val dictionary = Dictionary.loadModel(args, buffer)

            val quantInput = buffer.readUnsignedByte() != 0
            val quant_ = quantInput


            val input = if (quantInput) {
                loadQuantMatrix(buffer)
            } else {
                loadFloatArrayMatrixCPP(buffer)
            }
//
            if (!quantInput && dictionary.isPruned()) {
                throw RuntimeException("Invalid model file.\n"
                        + "Please download the updated model from www.fasttext.cc.\n"
                        + "See issue #332 on Github for more information.\n")
            }

            val qout = buffer.readUnsignedByte() != 0

            val output = if (quantInput && qout) {
                loadQuantMatrix(buffer)
            } else {
                loadFloatArrayMatrixCPP(buffer)
            }

            val loss = createLoss(args, output, args.model, dictionary)

            val normalizeGradient = args.model == ModelName.sup

            return FastText(args, dictionary, Model(input, output, loss, normalizeGradient), quant_)
        }
    }

    /**
     * Load binary model file. 这个二进制版本是C语言版本的模型
     * @param modelPath
     * @return FastTextModel
     * @throws Exception
     */
    @Throws(Exception::class)
    fun load(modelFile: File): FastText {

        if (!(modelFile.exists() && modelFile.isFile && modelFile.canRead())) {
            throw IOException("Model file cannot be opened for loading!")
        }

        return loadCModel(modelFile.inputStream())
    }

}